Message details view-models

BarAgam opened the pull request:
Created message details view-models and instantiated the different kinds at GGApplication for future use

barmako committed:
Message-detail view-model completed.
Create one VM for each message type, with a common abstract VM that exposes shared message data.
 Each VM will raise dataChanged only when a new compatible type message is selected.

barmako committed:
Instantiated and exposed messages view-models at GGApplication

review-ninja commented:
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/60" target="_blank"><img src="https://app.review.ninja/52010471/pull/60/badge" alt="ReviewNinja"/>

barmako committed:
Extracted functionality of selected-message-view-models into a separate class.
Added suiting test.

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/GGApplication.java:
```diff
@@ -27,6 +30,9 @@
     private MessagesReadStatusModel mMessagesReadStatusModel;
     private SelectedMessageModel mSelectedMessageModel;
```
consider create class for holding all models and view models

BarAgam replied:
The next task adds another field. 
I'll do it there

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/ImageMessageDetailViewModel.java:
```diff
@@ -0,0 +1,47 @@
+        super(selectedMessageModel);
+    }
+
+    public String getImageUrl() {
```
you missed here getSource(). !fix

BarAgam replied:
!fixed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/LatLongMessageDetailViewModel.java:
```diff
@@ -0,0 +1,29 @@
+        super(selectedMessageModel);
+    }
+
+    public PointGeometry getPointGeometry() {
```
consider change it's name to getLocation?

BarAgam replied:
but it returns a point geometry

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/MessageDetailViewModel.java:
```diff
@@ -0,0 +1,40 @@
+ * Verifies selected message is of appropriate type before each method call.
+ * Notifies observers on data changes only when selected message is of appropriate type.
+ */
+abstract class MessageDetailViewModel extends SelectedMessageViewModel implements DataChangedObservable {
```
why it is not public?
I would like to hold reference to MessageDetailViewModel so i would not need to update the title every time.

BarAgam replied:
Fine, you're right.
There, I said it.
Happy now ?

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    protected void validateSelectedMessage() {
+        if (!mSelectedMessageModel.isAnySelected()) {
```
isAnySelected is a useful method also for the views

BarAgam replied:
The container viewmodel wil expose such a method

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/ViewsModels/messages/SelectedMessageViewModel.java:
```diff
@@ -0,0 +1,59 @@
+
+    protected abstract boolean shouldNotifyObservers();
+
+    public static class NoSelectedMessageException extends RuntimeException {
```
if you want to control the flow with this exception make also the upper class public !resolve

BarAgam replied:
change it to have the default access modifier. !resolved

barmako committed:
CR fixes:
1. Exposed image-source on ImageMessageDetailViewModel (+tests)
2. Made MessageDetailViewModel public
3. Made custom exception related to illegal view-model requests access modifier to be default (so they wont be used outside the package)
